drafts/25-xxx-specifications.txt

```diff
@@ -567,11 +567,121 @@ nd20. The result of evaluating a processor-dependent directive is processor-depe
+| ID           | The expansion of the object-like macro ID   |
+| ID (args)    | The expansion of the function-like macro ID |
+| WHOLE_NUMBER | Decimal value of WHOLE_NUMBER               |
+| ( expr )     | Parenthesized expressions                   |
+|--------------+---------------------------------------------|
+
+
+
+5.1 Operators allowed in controlling expressions
+------------------------------------------------
+
+To maintain compatibility with the use of C preprocessing directives
+in many existing Fortran programs, the operators allowed in
+controlling expressions in #if and #elif expressions are a subset of
+those defined in the C 2023 standard §6.5 "Expressions" and §6.6
+"Constant expressions".
+
+A "precedence level" is assigned to each operator that determines
+how the operators combine with sub-expressions containing other operators
+at different precedence levels.
+
+An "associativity" is assigned to each operator that determines how
+operators at the same precedence level are combined.
+    - "left" means the operator binds to the left,
+    - "right" means the operator binds to the right.
+    - "nonassoc" means that the operator is not associative.
+
+The following table describes the semantics of the allowed operators
+in conditional expressions. The table is grouped by precedence level,
+from lowest precedence to highest.
+
+We label subexpressions "e1", "e2", and "e3" to aid in describing the
+evaluation semantics. Unless otherwise specified, all operators
+evaluate with the same semantics as their Fortran counterparts.
+
+
+| Prec |   Op    |    Syntax    |  Assoc'y |  Evaluation Semantics      |
+|------+---------+--------------+----------+----------------------------|
+| low  |   ? :   | e1 ? e2 : e3 |  right   |      conditional-expr      |
+|------+---------+--------------+----------+----------------------------|
+|      |   ¦¦    |   e1 || e2   |   left   |       Fortran .OR.         |
+|------+---------+--------------+----------+----------------------------|
+|      |   &&    |   e1 && e2   |   left   |       Fortran .AND.        |
+|------+---------+--------------+----------+----------------------------|
+|      |    ¦    |   e1 | e2    |   left   |    Fortran IOR(e1, e2)     |
+|      |    ^    |   e1 ^ e2    |   left   |    Fortran IAND(e1, e2)    |
+|      |    &    |   e1 & e2    |   left   |    Fortran IEOR(e1, e2)    |
+|------+---------+--------------+----------+----------------------------|
+|      |   ==    |   e1 == e2   |   left   | 1 if e1 == e2, 0 otherwise |
+|      |   !=    |   e1 != e2   |   left   | 1 if e1 /= e2, 0 otherwise |
+|------+---------+--------------+----------+----------------------------|
+|      |    >    |   e1 > e2    | nonassoc | 1 if e1 > e2, 0 otherwise  |
+|      |   >=    |   e1 >= e2   | nonassoc | 1 if e1 >= e2, 0 otherwise |
+|      |    <    |   e1 < e2    | nonassoc | 1 if e1 < e2, 0 otherwise  |
+|      |   <=    |   e1 <= e2   | nonassoc | 1 if e1 <= e2, 0 otherwise |
```

Comment:
In C (and hence CPP) _relational-expression_ operators are left associative.

Example:
```
#if 0 > -4 > 0
left-assoc
#else
right-assoc
#endif
```

Expands to "left-assoc" in CPP.

However it appears some FPPs may (unintentionally?) [diverge on this detail](https://godbolt.org/z/Yfzc4hnr5). Another great example why we need standardization!

```suggestion
|      |    >    |   e1 > e2    |   left   | 1 if e1 > e2, 0 otherwise  |
|      |   >=    |   e1 >= e2   |   left   | 1 if e1 >= e2, 0 otherwise |
|      |    <    |   e1 < e2    |   left  | 1 if e1 < e2, 0 otherwise  |
|      |   <=    |   e1 <= e2   |   left  | 1 if e1 <= e2, 0 otherwise |
```

Reply:
I put them as non-associative in the Fortran preprocessor, for potential future compatibility if we ever add Fortran operators. The Fortran operators are non-associative, so we might want to disallow expressions like `0 > -4 > 0` now. (And I think chaining relational operators is terrible, but that's just my opinion. If I saw that in C code myself, I would probably replace it with something people could immediately understand.)